napari/utils/chunk/_loader.py

```diff
@@ -0,0 +1,480 @@
+   multiple pools would result in more workers than we want.
+
+Think of the ChunkLoader as a shared resource like "the filesystem" where
+multiple clients can be access it at the same time, but it is the interface
+to just one physical resource.
+"""
+import logging
+from concurrent import futures
+from contextlib import contextmanager
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+from ...types import ArrayLike
+from ...utils.event import EmitterGroup
+from ..perf import perf_timer
+from ._cache import ChunkCache
+from ._config import async_config
+from ._delay_queue import DelayQueue
+from ._info import LayerInfo, LoadType
+from ._request import ChunkKey, ChunkRequest
+
+LOGGER = logging.getLogger("ChunkLoader")
+
+# Executor for either a thread pool or a process pool.
+PoolExecutor = Union[futures.ThreadPoolExecutor, futures.ProcessPoolExecutor]
+
+
+def _chunk_loader_worker(request: ChunkRequest) -> ChunkRequest:
+    """This is the worker thread or process that loads the array.
+
+    We call np.asarray() in a worker because it might lead to IO or
+    computation which would block the GUI thread.
+
+    Parameters
+    ----------
+    request : ChunkRequest
+        The request to load.
+    """
+    request.load_chunks()  # loads all chunks in the request
+    return request
+
+
+def _create_executor(use_processes: bool, num_workers: int) -> PoolExecutor:
+    """Return the thread or process pool executor.
+
+    Parameters
+    ----------
+    use_processes : bool
+        If True use processes, otherwise threads.
+    num_workers : int
+        The number of worker threads or processes.
+    """
+    if use_processes:
+        return futures.ProcessPoolExecutor(max_workers=num_workers)
+    return futures.ThreadPoolExecutor(max_workers=num_workers)
+
+
+class ChunkLoader:
+    """Loads chunks in worker threads or processes.
+
+    We cannot call np.asarray() in the GUI thread because it might block on
+    IO or a computation. So the ChunkLoader calls np.asarray() for us in
+    a worker thread or process.
+
+    Attributes
+    ----------
+    synchronous : bool
+        If True LoadType.AUTO layers are loading synchronously in the GUI thread.
+    load_seconds : float
+        Sleep this long in the worker during a load.
+    executor : Union[ThreadPoolExecutor, ProcessPoolExecutor]
+        Our thread or process pool executor.
+    futures : FutureMap
+        In progress futures for each layer (data_id).
+    cache : ChunkCache
+        Cache of previously loaded chunks.
+    events : EmitterGroup
+        We only signal one event: chunk_loaded.
+    """
+
+    FutureMap = Dict[int, List[futures.Future]]
+    LayerMap = Dict[int, LayerInfo]
+
+    def __init__(self):
+        self.synchronous: bool = async_config.synchronous
+        self.load_seconds: float = async_config.load_seconds
+
+        use_processes: bool = async_config.use_processes
+        num_workers: int = async_config.num_workers
```

Comment:
it appears that the global `async_config` is pretty much exclusively used to instantiate the `ChunkLoader`... which is also a global singleton.  Curious why you chose to design it that way, rather than just having `ChunkLoader` initialize in the regular way (it could still of course look for an env var and a config file if).  The only other place that `async_config` is used is in the commands, which could also just look at the global `chunk_loader` object instead.

Reply:
I'm confused by what "the regular way" is. The config files is an async config file, it's not a chunk loader config file, is that what you mean? For instance it will need to configural graphics properties related to async that have nothing to do with the Chunkloader? Right? Async is not just the chunkloader, it's a bigger feature that related to loading but also graphics?